### PR TITLE
Add tests for AddParticipantServlet and SearchMatchServlet and add helper classes

### PR DESCRIPTION
**Pull request by jzwang8**
Files changed/added:
- AddParticipantServlet: moved all functionality to AddParticipantHelper 
- AddParticipantHelper: implement add participant logic
- AddParticipantHelperTest: add tests for helper class
- SearchMatchServlet: moved all functionality to SearchMatchServlet
- SearchMatchHelper: implement search match logic
- SearchMatchHelperTest: add tests for helper class
- MatchDatastore/ParticipantDatastore/UserDatastore: make not final for mocking
- MatchPreference: add forStringValue method

### CODE REVIEW

File: backend/src/main/java/com/google/sps/data/MatchPreference.java [hunk 1]
```diff
@@ -41,4 +41,18 @@ public static MatchPreference forIntValue(int value) {
...
+      case "different":
+        return MatchPreference.DIFFERENT;
+      case "any":
+        return MatchPreference.ANY;
+      case "similar":
+        return MatchPreference.SIMILAR;
+      default:
+        throw new IllegalStateException("Unknown enum value.");
+    }
+  }
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/data/MatchPreference.java:57]: This was submitted already in another PR right?

File: backend/src/main/java/com/google/sps/servlets/AddParticipantServlet.java [hunk 2]
```diff
@@ -16,171 +16,38 @@
...
-      default:
-        return null;
-    }
+    AddParticipantHelper addParticipantHelper =
+        new AddParticipantHelper(
+            Clock.systemUTC(),
+            matchDatastore,
+            participantDatastore,
+            userDatastore,
+            usernameService);
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/servlets/AddParticipantServlet.java:50]: This can be reused if it becomes a field on this class.

File: backend/src/main/java/com/google/sps/servlets/SearchMatchServlet.java [hunk 3]
```diff
@@ -16,183 +16,31 @@
...
-  /** Send email notification when participant is expired */
-  private void sendExpiredEmail() {
-    try {
-      emailNotifier = createOrGetEmailNotifier();
-      emailNotifier.sendExpiredEmail(getUsername(), userService.getCurrentUser().getEmail());
-    } catch (GeneralSecurityException | MessagingException | IOException e) {
-      e.printStackTrace();
-    }
+    SearchMatchHelper searchMatchHelper =
+        new SearchMatchHelper(matchDatastore, participantDatastore, usernameService);
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/servlets/SearchMatchServlet.java:43]: make this a field?

File: backend/src/test/java/com/google/sps/AddParticipantHelperTest.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+  /** Mock HttpServletRequest method */
+  private BufferedReader getReader(JSONObject jsonObject) {
+    return new BufferedReader(new StringReader(jsonObject.toString()));
+  }
+
+  /** Mock HttpServletResponse method */
+  private PrintWriter getWriter() {
+    return new PrintWriter(System.out);
+  }
```
Comment (kmh287) [inline backend/src/test/java/com/google/sps/AddParticipantHelperTest.java:347]: WDYT about inlining this?

File: backend/src/test/java/com/google/sps/SearchMatchHelperTest.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        new SearchMatchHelper(matchDatastore, participantDatastore, usernameService);
+    searchMatchHelper.doGet(request, response);
+
+    verify(response).setStatus(HttpServletResponse.SC_OK, "Participant has a match!");
+  }
+
+  /** Mock HttpServletResponse method */
+  private PrintWriter getWriter() {
+    return new PrintWriter(System.out);
+  }
```
Comment (kmh287) [inline backend/src/test/java/com/google/sps/SearchMatchHelperTest.java:207]: inline?

File: backend/src/test/java/com/google/sps/SearchMatchHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void noParticipantInDatastore() throws IOException {
+    searchMatchHelper =
+        new SearchMatchHelper(matchDatastore, participantDatastore, usernameService);
```
Comment (kmh287) [inline backend/src/test/java/com/google/sps/SearchMatchHelperTest.java:83]: Since this happens in all the tests, can it move into the setUp method?

File: backend/src/main/java/com/google/sps/data/MatchPreference.java [hunk 7]
```diff
@@ -41,4 +41,18 @@ public static MatchPreference forIntValue(int value) {
     }
     throw new IllegalStateException("Unknown enum value.");
   }
+
+  /** Parse match preference string to MatchPreference */
+  public static MatchPreference forStringValue(String value) {
+    switch (value.toLowerCase()) {
```
Comment (TcgVanguardTroll) [inline backend/src/main/java/com/google/sps/data/MatchPreference.java:47]: NIt : do you think the lower case value should be made into a variable so that the function doesn't have to be called each time

File: backend/src/test/java/com/google/sps/SearchMatchHelperTest.java (hunk 6 above)
Comment (jzwang8) [reply backend/src/test/java/com/google/sps/SearchMatchHelperTest.java:83]: Even though it happens in all the tests, 4/5 tests don't actually use the mock datastores since testing the different paths through the servlet requires actually putting entities into the datastore.

File: backend/src/main/java/com/google/sps/data/MatchPreference.java (hunk 1 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/data/MatchPreference.java:57]: Yes but I needed it for this PR so I just added it so it would build.

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    if (obj == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Could not read request body");
+      return;
+    }
+    JSONObject formDetails = obj.getJSONObject(REQUEST_FORM_DETAILS);
+
+    // Get new Participant from input parameters
+    Participant newParticipant = getParticipantFromInputs(response, formDetails);
+    if (newParticipant == null) {
+      return;
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:86]: We'll want to send a response back to the client here

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    return new Participant(
+        username,
+        startTimeAvailable,
+        endTimeAvailable,
+        duration,
+        role,
+        productArea,
+        interests,
+        matchPreference,
+        /* matchId=*/ 0,
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:178]: tiny nit: space after the = so that it's `/* matchID= */`

File: backend/src/main/java/com/google/sps/data/MatchPreference.java [hunk 10]
```diff
@@ -42,16 +42,18 @@ public static MatchPreference forIntValue(int value) {
...
+    switch (lowerCasePreference) {
       case "different":
         return MatchPreference.DIFFERENT;
       case "any":
         return MatchPreference.ANY;
       case "similar":
         return MatchPreference.SIMILAR;
       default:
-        throw new IllegalStateException("Unknown enum String value.");
+        throw new IllegalStateException("Unknown enum value.");
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/data/MatchPreference.java:56]: while we're here, let's concatenate `value` to the exception reason. That way if we ever get an unexpected value we know what it is and it'll be easier to debug.

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+      System.out.println("match not found");
+    }
+
+    // Confirm received form input
+    response.setContentType("text/plain;charset=UTF-8");
+    response.getWriter().println("Received form input details and queried!");
+  }
+
+  /** Retrieve JSON body payload and convert to a JSONObject for parsing purposes */
+  private JSONObject retrieveRequestBody(HttpServletRequest request) throws IOException {
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:125]: does this method still throw IOException or would any exception be caught below?

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+    StringBuilder requestBuffer = new StringBuilder();
+    try {
+      BufferedReader reader = request.getReader();
+      String currentLine;
+      while ((currentLine = reader.readLine()) != null) {
+        requestBuffer.append(currentLine);
+      }
+    } catch (IOException e) {
+      return null;
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:134]: Please add a Nullable annotation and a short description in the javadocs for when this would return null.

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    }
+    return new JSONObject(requestBuffer.toString());
+  }
+
+  /** Get a Participant from form inputs */
+  private Participant getParticipantFromInputs(HttpServletResponse response, JSONObject formDetails)
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:140]: Nullable + docs

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+  /** Retrieve user email address via Users API and parse for username */
+  public static String getUsername() {
+    UserService userService = UserServiceFactory.getUserService();
+    String email = userService.getCurrentUser().getEmail();
+    return email != null ? email.split("@")[0] : null;
+  }
+
+  /** Get list of Strings from JSONArray */
+  public static List<String> getListFromJsonArray(JSONArray array) {
```
Comment (kmh287) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:193]: private?

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+        MatchStatus.UNMATCHED,
+        timestamp);
+  }
+
+  /** Retrieve user email address via Users API and parse for username */
+  public static String getUsername() {
+    UserService userService = UserServiceFactory.getUserService();
+    String email = userService.getCurrentUser().getEmail();
+    return email != null ? email.split("@")[0] : null;
+  }
```
Comment (dhruvi618) [inline backend/src/main/java/com/google/sps/AddParticipantHelper.java:190]: Since the UsernameService class is in master now, WDYT about removing this and method and calling it via a UsernameService instead?

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java (hunk 14 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/AddParticipantHelper.java:193]: Actually, I don't need this method anymore because it's in UsernameService now

File: backend/src/main/java/com/google/sps/AddParticipantHelper.java (hunk 15 above)
Comment (jzwang8) [reply backend/src/main/java/com/google/sps/AddParticipantHelper.java]: Yes, it already uses UsernameService and never calls this getUsername(). I just forgot to remove it!
